protocols/pnet/src/crypt_writer.rs

```diff
@@ -0,0 +1,146 @@
+
+use futures::{
+    io::{self, AsyncWrite},
+    ready,
+    task::{Context, Poll},
+};
+use log::trace;
+use pin_project::pin_project;
+use salsa20::{stream_cipher::SyncStreamCipher, XSalsa20};
+use std::{fmt, pin::Pin};
+
+/// A writer that encrypts and forwards to an inner writer
+#[pin_project]
+pub struct CryptWriter<W> {
+    #[pin]
+    inner: W,
+    buf: Vec<u8>,
+    cipher: XSalsa20,
+}
+
+impl<W: AsyncWrite> CryptWriter<W> {
+    /// Creates a new `CryptWriter` with the specified buffer capacity.
+    pub fn with_capacity(capacity: usize, inner: W, cipher: XSalsa20) -> CryptWriter<W> {
+        CryptWriter {
+            inner,
+            buf: Vec::with_capacity(capacity),
+            cipher,
+        }
+    }
+
+    /// Gets a pinned mutable reference to the inner writer.
+    ///
+    /// It is inadvisable to directly write to the inner writer.
+    pub fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut W> {
+        self.project().inner
+    }
+}
+
+/// Write the contents of a Vec<u8> into an AsyncWrite.
+///
+/// The handling 0 byte progress and the Interrupted error was taken from BufWriter in async_std.
+///
+/// If this fn returns Ready(Ok(())), the buffer will be completely flushed.
+fn poll_flush_buf<W: AsyncWrite>(
+    inner: &mut Pin<&mut W>,
+    buf: &mut Vec<u8>,
+    cx: &mut Context<'_>,
+) -> Poll<io::Result<()>> {
+    let mut ret = Poll::Ready(Ok(()));
+    let mut written = 0;
+    let len = buf.len();
+    while written < len {
+        match inner.as_mut().poll_write(cx, &buf[written..]) {
+            Poll::Ready(Ok(n)) => {
+                if n > 0 {
+                    // we made progress, so try again
+                    written += n;
+                } else {
+                    // we got Ok but got no progress whatsoever, so bail out so we don't spin writing 0 bytes.
+                    ret = Poll::Ready(Err(io::Error::new(
+                        io::ErrorKind::WriteZero,
+                        "Failed to write buffered data",
+                    )));
+                    break;
+                }
+            }
+            Poll::Ready(Err(e)) => {
+                // Interrupted is the only error that we consider to be recoverable by trying again
+                if e.kind() != io::ErrorKind::Interrupted {
+                    // for any other error, don't try again
+                    ret = Poll::Ready(Err(e));
+                    break;
+                }
+            }
+            Poll::Pending => break,
+        }
+    }
+    if written > 0 {
+        buf.drain(..written);
+    }
+    ret
+}
+
+impl<W: AsyncWrite> AsyncWrite for CryptWriter<W> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut this = self.project();
+        // completely flush the buffer, returning pending if not possible
+        ready!(poll_flush_buf(&mut this.inner, this.buf, cx))?;
```

Comment:
Ideally should only be done if there isn't enough space to fit all the data.

Reply:
Here is my thinking behind the "just do encryption" decision: the private swarms are an optional feature that is sometimes, but not always, enabled. If buffering is necessary it should be done somewhere else so we also have it when private swarms are disabled. If buffering is happening at a higher layer, data will be sent to / read from this layer already in large chunks, so no need for additional chunking.